src/libopensc/card-cac.c

```diff
@@ -0,0 +1,1580 @@
+		r = *recvbuflen;
+	}
+
+err:
+	SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, r);
+}
+
+/*
+ * Read a CAC TLV file. Parameters specify if the TLV file is TL (Tag/Length) file or a V (value) file
+ */
+#define HIGH_BYTE_OF_SHORT(x) (((x)>> 8) & 0xff)
+#define LOW_BYTE_OF_SHORT(x) ((x) & 0xff)
+static int cac_read_file(sc_card_t *card, int file_type, u8 **out_buf, size_t *out_len)
+{
+	u8 params[2];
+	u8 count[2];
+	u8 *out = NULL;
+	u8 *out_ptr;
+	size_t offset = 0;
+	size_t size = 0;
+	size_t left = 0;
+	size_t len;
+	int r;
+
+	params[0] = file_type;
+	params[1] = 2;
+
+	/* get the size */
+	len = sizeof(count);
+	out_ptr = count;
+	r = cac_apdu_io(card, CAC_INS_READ_FILE, 0, 0, &params[0], sizeof(params), &out_ptr, &len);
+	if (r < 0)
+		goto fail;
+
+	left = size = lebytes2ushort(count);
+sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,"got %d bytes out_ptr=%lx count&=%lx count[0]=0x%02x count[1]=0x%02x, len=0x%04x (%d)", len, (unsigned long) out_ptr, (unsigned long)&count, count[0], count[1], size, size); 
+	out = out_ptr = malloc(size);
+	if (out == NULL) {
+		r = SC_ERROR_OUT_OF_MEMORY;
+		goto fail;
+	}
+	for (offset += 2; left > 0; offset += len, left -= len, out_ptr += len) {
+		len = MIN(left, CAC_MAX_CHUNK_SIZE);
+		params[1] = len;
+		r = cac_apdu_io(card, CAC_INS_READ_FILE, HIGH_BYTE_OF_SHORT(offset), LOW_BYTE_OF_SHORT(offset), 
+						&params[0], sizeof(params), &out_ptr, &len);
+		if (r < 0) {
+			goto fail;
+		}
+	}
+	*out_len = size;
+	*out_buf = out;
+	return SC_SUCCESS;
+fail:
+	if (out) 
+		free(out);
+	*out_len = 0;
+	return r;
+}
+
+/*
+ * OLD cac read certificate, only use with CAC-1 card.
+ */
+static int cac_cac1_get_certificate(sc_card_t *card, u8 **out_buf, size_t *out_len)
+{
+	u8 buf[CAC_MAX_SIZE];
+	u8 *out_ptr;
+	size_t size = 0;
+	size_t left = 0;
+	size_t len, next_len;
+	sc_apdu_t apdu;
+	int r;
+
+
+	/* get the size */
+	size = left = *out_buf ? *out_len : sizeof(buf);
+	out_ptr = *out_buf ? *out_buf : buf;
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, CAC_INS_GET_CERTIFICATE, 0, 0 );
+	next_len = MIN(left, 100);
+	for (; left > 0; left -= len, out_ptr += len) {
+		len = next_len;
+		apdu.resp = out_ptr;
+		apdu.le = len;
+		apdu.resplen = left;
+
+		r = sc_transmit_apdu(card, &apdu);
+		if (r < 0) {
+		 	break;
+		}
+		/* in the old CAC-1, 0x63 means 'more data' in addition to 'pin failed' */
+		if (apdu.sw1 != 0x63)  {
+			/* we've either finished reading, or hit an error, break */
+			r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+			left -= len;
+			break;
+		}
+		next_len = MIN(left,apdu.sw2);
+	}
+	if (r < 0) {
+		return r;
+	}
+	r = size - left;
+	if (*out_buf == NULL) {
+		*out_buf = malloc(r);
+		if (*out_buf == NULL) {
+			return SC_ERROR_OUT_OF_MEMORY;
+		}
+		memcpy(*out_buf, buf, r);
+	}
+	*out_len = r;
+	return r;
+}
+
+/* create a fake tag/length file for cac1 cards based on the val_len */
+static int cac_cac1_get_cert_tag(sc_card_t *card, size_t val_len, u8 **tlp, size_t *tl_len_p) 
```

Comment:
The above encoding is `simple TLV`, unlike the BER/DER encoding used in `asn1.h` and therefore I am not able to simplify the encoding using the above library. Though it might be possible to use it in the decoding.

Example of difference between BER TLV and Simple TLV can be found in an Appendix C of [CAC End-Point Implementation Guide](http://www.cac.mil/docs/ref1.c.i-CAC_End-Point_Implementation_Guide_v1-22.pdf) (linked from previous comments).